app/src/main/java/com/github/multimatum_team/multimatum/activity/GroupDetailsActivity.kt

```diff
@@ -0,0 +1,213 @@
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_group_details)
+
+        // Get the ID of the group from intent extras
+        groupID = intent.getStringExtra(EXTRA_ID) as GroupID
+
+        // Get the necessary widgets
+        groupNameView = findViewById(R.id.group_details_name)
+        groupOwnerView = findViewById(R.id.group_details_owner)
+        groupMembersView = findViewById(R.id.group_details_members)
+
+        groupInviteButton = findViewById(R.id.group_details_invite_button)
+        groupDeleteOrLeaveButton = findViewById(R.id.group_details_delete_or_leave_button)
+
+        // Initialize UI widgets
+        initTextInput()
+        initGroupMemberView()
+        initLeaveButton()
+    }
+
+    /**
+     * Update view when remove data changed.
+     */
+    private fun updateView() {
+        groupNameView.focusable = if (currentUserIsGroupOwner()) {
+            View.FOCUSABLE_AUTO
+        } else {
+            View.NOT_FOCUSABLE
+        }
+
+        groupNameView.text = SpannableStringBuilder.valueOf(group.name)
+        val ownerName = runBlocking { userRepository.fetch(group.owner).name }
+        groupOwnerView.text = getString(R.string.group_owner, ownerName)
+
+        if (currentUserIsGroupOwner()) {
+            initDeleteButton()
+        } else {
+            initLeaveButton()
+        }
+    }
+
+    /**
+     * Returns whether the group we are seeing is owned by the authenticated user.
+     */
+    private fun currentUserIsGroupOwner(): Boolean =
+        group.owner == authViewModel.getUser().value!!.id
+
+    /**
+     * Setup the text input
+     */
+    private fun initTextInput() {
+        // Adding a listener to handle the "DONE" key pressed.
+        groupNameView.setOnIMEActionDone(this) { newName ->
+            groupViewModel.renameGroup(groupID, newName)
+        }
+    }
+
+    /**
+     * Initialize group member view, setting its adapter and observing the group repository when
+     * remote data changes.
+     */
+    private fun initGroupMemberView() {
+        val adapter =
+            GroupMemberAdapter(this, userRepository, authViewModel, groupViewModel)
+
+        groupMembersView.adapter = adapter
+        groupMembersView.layoutManager = LinearLayoutManager(this)
+
+        groupViewModel.getGroups().observe(this) { groups ->
+            when (groups[groupID]) {
+                null -> finish()
+                else -> {
+                    group = groups[groupID]!!
+                    adapter.setGroup(group)
+                    updateView()
+                }
+            }
+        }
+    }
+
+    /**
+     * Initialize groupDeleteOrLeaveButton with the right text and leave the group when clicked.
+     */
+    private fun initLeaveButton() {
+        groupDeleteOrLeaveButton.text = getString(R.string.group_leave)
+        groupDeleteOrLeaveButton.setOnClickListener {
```

Comment:
This could be setup with the layout. Is there a reason for this specific way ?

Reply:
You mean with two overlapping buttons and toggling which one is shown?

Reply:
I mean that you could link a function to a button via the .xml

Reply:
The issue is that both the text button and action on click can change, if I bind the action to a function that leaves or delete the group depending on whether the user owns the group, then it could be that the text and the action are out of sync